{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "78e6df46-8ed3-4174-87af-49eff95fd66d",
   "metadata": {},
   "source": [
    "# Preparing count matrices and variable genes for cNMF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "649b75e7-98bd-4081-9c6a-3b4aa429794e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(KernSmooth)\n",
    "library(reshape2)\n",
    "library(Rtsne)\n",
    "\n",
    "library(ggplot2)\n",
    "library(RColorBrewer)\n",
    "library(gplots)\n",
    "library(ggrepel)\n",
    "library(dplyr)\n",
    "library(plyr)\n",
    "library(DescTools)\n",
    "\n",
    "library(Seurat)\n",
    "library(matrixStats)\n",
    "library(tidyr)\n",
    "\n",
    "library(dbscan)\n",
    "\n",
    "\n",
    "source(\"../sc_analysis_palettes_and_helpers.r\")\n",
    "\n",
    "#wd <- '/local/users/nfresma/'\n",
    "\n",
    "wd <- './nmf_data/'\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21be7593-fe15-468a-a914-f2f261dc13fe",
   "metadata": {},
   "source": [
    "## Write count matrixes and variable genes lists for various sets of NB cell datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38de631c-13e4-485b-9173-cae8e4a9fc07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load NB cell Seurat object\n",
    "\n",
    "dat <- readRDS('seur_all_tumors_NB_cells.rds')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da2dc5bb-ffdc-417c-98ac-735f808bdb63",
   "metadata": {},
   "source": [
    "### For entire dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8ff2c96a-5f93-4036-82a4-6d416243676b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating cluster mycn_tum_01\n",
      "\n",
      "Calculating cluster mycn_tum_02\n",
      "\n",
      "Calculating cluster mycnlmo1_ventral_tum\n",
      "\n",
      "Calculating cluster mycnlmo1_kidney_tum\n",
      "\n",
      "Calculating cluster healthyhk_ctrl\n",
      "\n",
      "Calculating cluster lmo1_ctrl\n",
      "\n",
      "Calculating cluster mycn_tum_04\n",
      "\n",
      "Calculating cluster multi_seq_1_dr_tum\n",
      "\n",
      "Calculating cluster multi_seq_2\n",
      "\n",
      "Calculating cluster multi_seq_3\n",
      "\n",
      "Calculating cluster multi_4_allo1\n",
      "\n",
      "Calculating cluster multi_allos_1_tum\n",
      "\n",
      "Calculating cluster multi_bAllos_1_tum\n",
      "\n",
      "Calculating cluster multi_bAllos_2_tum\n",
      "\n",
      "Calculating cluster multi_bAllos_3_tum\n",
      "\n",
      "Calculating cluster multi_bAllos_4_tum\n",
      "\n",
      "Calculating cluster multiseq_17\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'atad2'</li><li>'blm'</li><li>'brip1'</li><li>'casp8ap2'</li><li>'cdc45'</li><li>'cdc6'</li><li>'cdca7a'</li><li>'chaf1b'</li><li>'dscc1'</li><li>'dtl'</li><li>'e2f8'</li><li>'exo1'</li><li>'fen1'</li><li>'gins2'</li><li>'gmnn'</li><li>'hells'</li><li>'mcm2'</li><li>'mcm4'</li><li>'mcm5'</li><li>'mcm6'</li><li>'msh2'</li><li>'nasp'</li><li>'pcna'</li><li>'pola1'</li><li>'pold3'</li><li>'prim1'</li><li>'rad51'</li><li>'rad51ap1'</li><li>'rfc2'</li><li>'rpa2'</li><li>'rrm1'</li><li>'rrm2'</li><li>'slbp'</li><li>'tipin'</li><li>'tyms'</li><li>'ubr7'</li><li>'uhrf1'</li><li>'unga'</li><li>'ungb'</li><li>'usp1'</li><li>'wdr76'</li><li>'anln'</li><li>'anp32e'</li><li>'aurka'</li><li>'birc5a'</li><li>'birc5b'</li><li>'bub1'</li><li>'cbx5'</li><li>'ccnb2'</li><li>'cdc20'</li><li>'cdca8'</li><li>'cdk1'</li><li>'cenpe'</li><li>'cenpf'</li><li>'ckap2l'</li><li>'ckap5'</li><li>'cks1b'</li><li>'cks2'</li><li>'ctcf'</li><li>'dlgap5'</li><li>'ect2'</li><li>'g2e3'</li><li>'gas2l3'</li><li>'gtse1'</li><li>'hmgb2a'</li><li>'hmgb2b'</li><li>'hmmr'</li><li>'kif11'</li><li>'kif20ba'</li><li>'kif20bb'</li><li>'kif23'</li><li>'kif2c'</li><li>'lbr'</li><li>'mki67'</li><li>'ncapd2'</li><li>'ndc80'</li><li>'nek2'</li><li>'nuf2'</li><li>'nusap1'</li><li>'rangap1a'</li><li>'rangap1b'</li><li>'tacc3'</li><li>'tmpoa'</li><li>'tmpob'</li><li>'top2a'</li><li>'tpx2'</li><li>'ttk'</li><li>'tubb4b'</li><li>'ube2c'</li><li>'stmn1a'</li><li>'mki67'</li><li>'plk1'</li><li>'top2a'</li><li>'nusap1'</li><li>'ccnb1'</li><li>'ube2c'</li><li>'aspm'</li><li>'tpx2'</li><li>'kifc1'</li><li>'smc2'</li><li>'hmgb2a'</li><li>'hmgn2'</li><li>'DUT'</li><li>'tubb2b'</li><li>'tuba8l4'</li><li>'stmn1b'</li><li>'pcna'</li><li>'hmgb2b'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'atad2'\n",
       "\\item 'blm'\n",
       "\\item 'brip1'\n",
       "\\item 'casp8ap2'\n",
       "\\item 'cdc45'\n",
       "\\item 'cdc6'\n",
       "\\item 'cdca7a'\n",
       "\\item 'chaf1b'\n",
       "\\item 'dscc1'\n",
       "\\item 'dtl'\n",
       "\\item 'e2f8'\n",
       "\\item 'exo1'\n",
       "\\item 'fen1'\n",
       "\\item 'gins2'\n",
       "\\item 'gmnn'\n",
       "\\item 'hells'\n",
       "\\item 'mcm2'\n",
       "\\item 'mcm4'\n",
       "\\item 'mcm5'\n",
       "\\item 'mcm6'\n",
       "\\item 'msh2'\n",
       "\\item 'nasp'\n",
       "\\item 'pcna'\n",
       "\\item 'pola1'\n",
       "\\item 'pold3'\n",
       "\\item 'prim1'\n",
       "\\item 'rad51'\n",
       "\\item 'rad51ap1'\n",
       "\\item 'rfc2'\n",
       "\\item 'rpa2'\n",
       "\\item 'rrm1'\n",
       "\\item 'rrm2'\n",
       "\\item 'slbp'\n",
       "\\item 'tipin'\n",
       "\\item 'tyms'\n",
       "\\item 'ubr7'\n",
       "\\item 'uhrf1'\n",
       "\\item 'unga'\n",
       "\\item 'ungb'\n",
       "\\item 'usp1'\n",
       "\\item 'wdr76'\n",
       "\\item 'anln'\n",
       "\\item 'anp32e'\n",
       "\\item 'aurka'\n",
       "\\item 'birc5a'\n",
       "\\item 'birc5b'\n",
       "\\item 'bub1'\n",
       "\\item 'cbx5'\n",
       "\\item 'ccnb2'\n",
       "\\item 'cdc20'\n",
       "\\item 'cdca8'\n",
       "\\item 'cdk1'\n",
       "\\item 'cenpe'\n",
       "\\item 'cenpf'\n",
       "\\item 'ckap2l'\n",
       "\\item 'ckap5'\n",
       "\\item 'cks1b'\n",
       "\\item 'cks2'\n",
       "\\item 'ctcf'\n",
       "\\item 'dlgap5'\n",
       "\\item 'ect2'\n",
       "\\item 'g2e3'\n",
       "\\item 'gas2l3'\n",
       "\\item 'gtse1'\n",
       "\\item 'hmgb2a'\n",
       "\\item 'hmgb2b'\n",
       "\\item 'hmmr'\n",
       "\\item 'kif11'\n",
       "\\item 'kif20ba'\n",
       "\\item 'kif20bb'\n",
       "\\item 'kif23'\n",
       "\\item 'kif2c'\n",
       "\\item 'lbr'\n",
       "\\item 'mki67'\n",
       "\\item 'ncapd2'\n",
       "\\item 'ndc80'\n",
       "\\item 'nek2'\n",
       "\\item 'nuf2'\n",
       "\\item 'nusap1'\n",
       "\\item 'rangap1a'\n",
       "\\item 'rangap1b'\n",
       "\\item 'tacc3'\n",
       "\\item 'tmpoa'\n",
       "\\item 'tmpob'\n",
       "\\item 'top2a'\n",
       "\\item 'tpx2'\n",
       "\\item 'ttk'\n",
       "\\item 'tubb4b'\n",
       "\\item 'ube2c'\n",
       "\\item 'stmn1a'\n",
       "\\item 'mki67'\n",
       "\\item 'plk1'\n",
       "\\item 'top2a'\n",
       "\\item 'nusap1'\n",
       "\\item 'ccnb1'\n",
       "\\item 'ube2c'\n",
       "\\item 'aspm'\n",
       "\\item 'tpx2'\n",
       "\\item 'kifc1'\n",
       "\\item 'smc2'\n",
       "\\item 'hmgb2a'\n",
       "\\item 'hmgn2'\n",
       "\\item 'DUT'\n",
       "\\item 'tubb2b'\n",
       "\\item 'tuba8l4'\n",
       "\\item 'stmn1b'\n",
       "\\item 'pcna'\n",
       "\\item 'hmgb2b'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'atad2'\n",
       "2. 'blm'\n",
       "3. 'brip1'\n",
       "4. 'casp8ap2'\n",
       "5. 'cdc45'\n",
       "6. 'cdc6'\n",
       "7. 'cdca7a'\n",
       "8. 'chaf1b'\n",
       "9. 'dscc1'\n",
       "10. 'dtl'\n",
       "11. 'e2f8'\n",
       "12. 'exo1'\n",
       "13. 'fen1'\n",
       "14. 'gins2'\n",
       "15. 'gmnn'\n",
       "16. 'hells'\n",
       "17. 'mcm2'\n",
       "18. 'mcm4'\n",
       "19. 'mcm5'\n",
       "20. 'mcm6'\n",
       "21. 'msh2'\n",
       "22. 'nasp'\n",
       "23. 'pcna'\n",
       "24. 'pola1'\n",
       "25. 'pold3'\n",
       "26. 'prim1'\n",
       "27. 'rad51'\n",
       "28. 'rad51ap1'\n",
       "29. 'rfc2'\n",
       "30. 'rpa2'\n",
       "31. 'rrm1'\n",
       "32. 'rrm2'\n",
       "33. 'slbp'\n",
       "34. 'tipin'\n",
       "35. 'tyms'\n",
       "36. 'ubr7'\n",
       "37. 'uhrf1'\n",
       "38. 'unga'\n",
       "39. 'ungb'\n",
       "40. 'usp1'\n",
       "41. 'wdr76'\n",
       "42. 'anln'\n",
       "43. 'anp32e'\n",
       "44. 'aurka'\n",
       "45. 'birc5a'\n",
       "46. 'birc5b'\n",
       "47. 'bub1'\n",
       "48. 'cbx5'\n",
       "49. 'ccnb2'\n",
       "50. 'cdc20'\n",
       "51. 'cdca8'\n",
       "52. 'cdk1'\n",
       "53. 'cenpe'\n",
       "54. 'cenpf'\n",
       "55. 'ckap2l'\n",
       "56. 'ckap5'\n",
       "57. 'cks1b'\n",
       "58. 'cks2'\n",
       "59. 'ctcf'\n",
       "60. 'dlgap5'\n",
       "61. 'ect2'\n",
       "62. 'g2e3'\n",
       "63. 'gas2l3'\n",
       "64. 'gtse1'\n",
       "65. 'hmgb2a'\n",
       "66. 'hmgb2b'\n",
       "67. 'hmmr'\n",
       "68. 'kif11'\n",
       "69. 'kif20ba'\n",
       "70. 'kif20bb'\n",
       "71. 'kif23'\n",
       "72. 'kif2c'\n",
       "73. 'lbr'\n",
       "74. 'mki67'\n",
       "75. 'ncapd2'\n",
       "76. 'ndc80'\n",
       "77. 'nek2'\n",
       "78. 'nuf2'\n",
       "79. 'nusap1'\n",
       "80. 'rangap1a'\n",
       "81. 'rangap1b'\n",
       "82. 'tacc3'\n",
       "83. 'tmpoa'\n",
       "84. 'tmpob'\n",
       "85. 'top2a'\n",
       "86. 'tpx2'\n",
       "87. 'ttk'\n",
       "88. 'tubb4b'\n",
       "89. 'ube2c'\n",
       "90. 'stmn1a'\n",
       "91. 'mki67'\n",
       "92. 'plk1'\n",
       "93. 'top2a'\n",
       "94. 'nusap1'\n",
       "95. 'ccnb1'\n",
       "96. 'ube2c'\n",
       "97. 'aspm'\n",
       "98. 'tpx2'\n",
       "99. 'kifc1'\n",
       "100. 'smc2'\n",
       "101. 'hmgb2a'\n",
       "102. 'hmgn2'\n",
       "103. 'DUT'\n",
       "104. 'tubb2b'\n",
       "105. 'tuba8l4'\n",
       "106. 'stmn1b'\n",
       "107. 'pcna'\n",
       "108. 'hmgb2b'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"atad2\"    \"blm\"      \"brip1\"    \"casp8ap2\" \"cdc45\"    \"cdc6\"    \n",
       "  [7] \"cdca7a\"   \"chaf1b\"   \"dscc1\"    \"dtl\"      \"e2f8\"     \"exo1\"    \n",
       " [13] \"fen1\"     \"gins2\"    \"gmnn\"     \"hells\"    \"mcm2\"     \"mcm4\"    \n",
       " [19] \"mcm5\"     \"mcm6\"     \"msh2\"     \"nasp\"     \"pcna\"     \"pola1\"   \n",
       " [25] \"pold3\"    \"prim1\"    \"rad51\"    \"rad51ap1\" \"rfc2\"     \"rpa2\"    \n",
       " [31] \"rrm1\"     \"rrm2\"     \"slbp\"     \"tipin\"    \"tyms\"     \"ubr7\"    \n",
       " [37] \"uhrf1\"    \"unga\"     \"ungb\"     \"usp1\"     \"wdr76\"    \"anln\"    \n",
       " [43] \"anp32e\"   \"aurka\"    \"birc5a\"   \"birc5b\"   \"bub1\"     \"cbx5\"    \n",
       " [49] \"ccnb2\"    \"cdc20\"    \"cdca8\"    \"cdk1\"     \"cenpe\"    \"cenpf\"   \n",
       " [55] \"ckap2l\"   \"ckap5\"    \"cks1b\"    \"cks2\"     \"ctcf\"     \"dlgap5\"  \n",
       " [61] \"ect2\"     \"g2e3\"     \"gas2l3\"   \"gtse1\"    \"hmgb2a\"   \"hmgb2b\"  \n",
       " [67] \"hmmr\"     \"kif11\"    \"kif20ba\"  \"kif20bb\"  \"kif23\"    \"kif2c\"   \n",
       " [73] \"lbr\"      \"mki67\"    \"ncapd2\"   \"ndc80\"    \"nek2\"     \"nuf2\"    \n",
       " [79] \"nusap1\"   \"rangap1a\" \"rangap1b\" \"tacc3\"    \"tmpoa\"    \"tmpob\"   \n",
       " [85] \"top2a\"    \"tpx2\"     \"ttk\"      \"tubb4b\"   \"ube2c\"    \"stmn1a\"  \n",
       " [91] \"mki67\"    \"plk1\"     \"top2a\"    \"nusap1\"   \"ccnb1\"    \"ube2c\"   \n",
       " [97] \"aspm\"     \"tpx2\"     \"kifc1\"    \"smc2\"     \"hmgb2a\"   \"hmgn2\"   \n",
       "[103] \"DUT\"      \"tubb2b\"   \"tuba8l4\"  \"stmn1b\"   \"pcna\"     \"hmgb2b\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Calculate variable genes and remove variable genes only found in certain datasets or specific to contaminating blood cells\n",
    "\n",
    "# Find genes that are differentially expressed between datasets to a certain cut-off\n",
    "Idents(dat) <- \"dataset\"\n",
    "\n",
    "dat <- FindVariableFeatures(object = dat,selection.method = 'vst', nfeatures = 2000)\n",
    "\n",
    "to.regress <- FindAllMarkers(object = dat, only.pos = TRUE, min.pct = 0.8, logfc.threshold = 0.3)\n",
    "to.regress <- to.regress[to.regress$cluster %like any% 'multi_%',]\n",
    "to.regress <- to.regress[to.regress$pct.1 > 0.9,]\n",
    "to.regress <- to.regress$gene\n",
    "\n",
    "# Load genes associated with the cell cycle that should also be excluded for and NMF.\n",
    "\n",
    "to.regress.cellcycle <- read.csv(\"../transcriptome_analysis/data/zfish_prolif_markers.csv\")\n",
    "to.regress.cellcycle <- as.character(to.regress.cellcycle$x)\n",
    "to.regress.cellcycle\n",
    "\n",
    "# Get genes specific to blood cell types that are contaminants in NB cell data to remove those from variable genes before clustering and NMF.\n",
    "regress_blood <- c(\"hbba1\",\"hbba1.1\",\"hbae1\",\"hbbe2\",\"hbaa1\",\"hbaa2\",\"hbba2\",\"hbbe1.1\",\n",
    "                   \"lyz\",\"lect2l\",\"mmp9\",\"npsn\",\"srgn\",\n",
    "                   \"cd74a\",\"cd74b\",\"dusp2\",\n",
    "                   \"c1qb\",\"c1qc\",\"grn1\",\"marco\",\n",
    "                   \"tox\",\"zap70\",\"sla2\",\"igic1s1\",\n",
    "                   \"nkl.1\",\"nkl.4\")\n",
    "\n",
    "\n",
    "var_feats <- VariableFeatures(object = dat)\n",
    "\n",
    "var_feats <- var_feats[!var_feats %in% c(\"mCherry\", \"EGFP\",\"MYCN\", \"LMO1\",\"dTomato\")] # Remove transgenes as they're not all present in all tumour samples.\n",
    "var_feats <- var_feats[!var_feats %in% to.regress]\n",
    "var_feats <- var_feats[!var_feats %in% to.regress.cellcycle]\n",
    "var_feats <- var_feats[!var_feats %in% regress_blood]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db8624d4-2a7e-423c-b09e-975d68c45c0e",
   "metadata": {},
   "source": [
    "Sets of genes excluded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e32153a8-3722-4227-9401-fdf86c5ec063",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"eef1b2\"            \"rpl10\"             \"rplp0\"            \n",
      " [4] \"tpt1\"              \"hsp90ab1\"          \"rps6\"             \n",
      " [7] \"rpl19\"             \"rack1\"             \"rpl17\"            \n",
      "[10] \"rps5\"              \"rpl9\"              \"rps8a\"            \n",
      "[13] \"rpl7\"              \"EGFP\"              \"rpl3\"             \n",
      "[16] \"rpl10a\"            \"FO704736.1\"        \"mibp2\"            \n",
      "[19] \"CR318588.1\"        \"CR318588.4\"        \"mt-atp6\"          \n",
      "[22] \"eef1db\"            \"actb2\"             \"rpl22l1\"          \n",
      "[25] \"rpl12\"             \"rps27.2\"           \"h2afx1\"           \n",
      "[28] \"nme2b.1\"           \"eef1g\"             \"eef2b\"            \n",
      "[31] \"rpl18a\"            \"rpl6\"              \"ppiaa\"            \n",
      "[34] \"rpl4\"              \"cfl1\"              \"b2m\"              \n",
      "[37] \"rplp1\"             \"si:ch211-5k11.8\"   \"NC-002333.17\"     \n",
      "[40] \"NC-002333.4\"       \"hbba1.1\"           \"rgs5a\"            \n",
      "[43] \"zgc:158463\"        \"CT027638.1\"        \"LMO1\"             \n",
      "[46] \"mCherry\"           \"rpl38\"             \"rpl22\"            \n",
      "[49] \"rpl36\"             \"rps28\"             \"rps29\"            \n",
      "[52] \"rplp2\"             \"hsph1\"             \"si:dkey-4p15.3\"   \n",
      "[55] \"DNAJA4\"            \"stip1\"             \"hsp90aa1.2\"       \n",
      "[58] \"mcl1a\"             \"dnajb1b\"           \"hspa5\"            \n",
      "[61] \"hsp70.2\"           \"hspa4a\"            \"hsp70.3\"          \n",
      "[64] \"si:ch211-131k2.3\"  \"uchl1\"             \"si:ch211-202a12.4\"\n",
      "[67] \"ubb\"               \"ccng1\"             \"hsp70l\"           \n",
      "[70] \"CR383676.1\"        \"hspa8\"             \"zgc:92066\"        \n",
      "[73] \"btf3\"              \"naca\"              \"tmsb\"             \n",
      "[76] \"rpl29\"             \"rpl39\"             \"ppiab\"            \n",
      "[79] \"rps19\"             \"zgc:114188\"        \"rpl36a\"           \n",
      "[82] \"atp5mc3b\"          \"tmsb4x\"            \"gapdhs\"           \n",
      "  [1] \"atad2\"    \"blm\"      \"brip1\"    \"casp8ap2\" \"cdc45\"    \"cdc6\"    \n",
      "  [7] \"cdca7a\"   \"chaf1b\"   \"dscc1\"    \"dtl\"      \"e2f8\"     \"exo1\"    \n",
      " [13] \"fen1\"     \"gins2\"    \"gmnn\"     \"hells\"    \"mcm2\"     \"mcm4\"    \n",
      " [19] \"mcm5\"     \"mcm6\"     \"msh2\"     \"nasp\"     \"pcna\"     \"pola1\"   \n",
      " [25] \"pold3\"    \"prim1\"    \"rad51\"    \"rad51ap1\" \"rfc2\"     \"rpa2\"    \n",
      " [31] \"rrm1\"     \"rrm2\"     \"slbp\"     \"tipin\"    \"tyms\"     \"ubr7\"    \n",
      " [37] \"uhrf1\"    \"unga\"     \"ungb\"     \"usp1\"     \"wdr76\"    \"anln\"    \n",
      " [43] \"anp32e\"   \"aurka\"    \"birc5a\"   \"birc5b\"   \"bub1\"     \"cbx5\"    \n",
      " [49] \"ccnb2\"    \"cdc20\"    \"cdca8\"    \"cdk1\"     \"cenpe\"    \"cenpf\"   \n",
      " [55] \"ckap2l\"   \"ckap5\"    \"cks1b\"    \"cks2\"     \"ctcf\"     \"dlgap5\"  \n",
      " [61] \"ect2\"     \"g2e3\"     \"gas2l3\"   \"gtse1\"    \"hmgb2a\"   \"hmgb2b\"  \n",
      " [67] \"hmmr\"     \"kif11\"    \"kif20ba\"  \"kif20bb\"  \"kif23\"    \"kif2c\"   \n",
      " [73] \"lbr\"      \"mki67\"    \"ncapd2\"   \"ndc80\"    \"nek2\"     \"nuf2\"    \n",
      " [79] \"nusap1\"   \"rangap1a\" \"rangap1b\" \"tacc3\"    \"tmpoa\"    \"tmpob\"   \n",
      " [85] \"top2a\"    \"tpx2\"     \"ttk\"      \"tubb4b\"   \"ube2c\"    \"stmn1a\"  \n",
      " [91] \"mki67\"    \"plk1\"     \"top2a\"    \"nusap1\"   \"ccnb1\"    \"ube2c\"   \n",
      " [97] \"aspm\"     \"tpx2\"     \"kifc1\"    \"smc2\"     \"hmgb2a\"   \"hmgn2\"   \n",
      "[103] \"DUT\"      \"tubb2b\"   \"tuba8l4\"  \"stmn1b\"   \"pcna\"     \"hmgb2b\"  \n"
     ]
    }
   ],
   "source": [
    "print(unique(to.regress))\n",
    "print(to.regress.cellcycle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3299970-334e-446a-a2e2-0f852d284b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get count matrix and remove cells that have zero counts of highly variable genes.\n",
    "# These will otherwise produce an error when running cNMF\n",
    "\n",
    "allcells_eset <- t(as.matrix(GetAssayData(dat, slot = 'counts')))\n",
    "\n",
    "cell_counts <- rowSums(allcells_eset[,colnames(allcells_eset) %in% var_feats])\n",
    "names(cell_counts) <- rownames(allcells_eset)\n",
    "\n",
    "cell_counts <- names(cell_counts[cell_counts == 0])\n",
    "allcells_eset <- allcells_eset[!rownames(allcells_eset) %in% cell_counts,]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8df64cf8-ab41-4e36-bd35-8cd1bf0bce35",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(var_feats, paste0(wd, \"./nmf_data/allmerged_nrblstm_varGenes.txt\"), quote = F, sep = \"\\t\", row.names = F, col.names = F)\n",
    "write.table(allcells_eset, paste0(wd, \"./nmf_data/allmerged_nrblstm_counts.tsv\", quote = F, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32e79a72-9af9-4cda-8429-54f6986d2359",
   "metadata": {},
   "source": [
    "## For MULTI-seq datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c0d1661-8fdd-42fc-a805-93d3ea03021d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load genes associated with the cell cycle that should also be excluded for clustering\n",
    "to.regress.cellcycle <- read.csv(\"../transcriptome_analysis/data/zfish_prolif_markers.csv\")\n",
    "to.regress.cellcycle <- as.character(to.regress.cellcycle$x)\n",
    "\n",
    "# Get genes specific to blood cell types that are contaminants in NB cell data\n",
    "regress_blood <- c(\"hbba1\",\"hbba1.1\",\"hbae1\",\"hbbe2\",\"hbaa1\",\"hbaa2\",\"hbba2\",\"hbbe1.1\",\n",
    "                   \"lyz\",\"lect2l\",\"mmp9\",\"npsn\",\"srgn\",\n",
    "                   \"cd74a\",\"cd74b\",\"dusp2\",\n",
    "                   \"c1qb\",\"c1qc\",\"grn1\",\"marco\",\n",
    "                   \"tox\",\"zap70\",\"sla2\",\"igic1s1\",\n",
    "                   \"nkl.1\",\"nkl.4\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c7443c-d286-49a7-bc9d-d37985eadfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Per dataset:\n",
    "# Subset full dataset -> Merge sets of cells derived from the same tumour samples but processed on separate 10x lanes.\n",
    "# Find variable genes.\n",
    "# Remove known cell cycle markers, contaminating blood / immune cell genes and transgenes from variable genes list.\n",
    "# Write out count matrix and variable genes.\n",
    "\n",
    "# MULTI-seq datasets to loop through\n",
    "datasets <- c('multi_seq_2_s1','multi_seq_3_s1','multi_4_allo1_1','multi_allos_1_tum',\n",
    "              'multi_bAllos_1_tum','multi_bAllos_2_tum','multi_bAllos_3_tum','multi_bAllos_4_tum','multiseq_17')\n",
    "\n",
    "Idents(dat) <- \"orig.ident\"\n",
    "\n",
    "for(i in 1:length(datasets)){\n",
    "    \n",
    "    subname <- datasets[i]\n",
    "    \n",
    "    if(subname == \"multi_seq_2_s1\"){\n",
    "        subdats <- c(\"multi_seq_2_s1\",\"multi_seq_2_s2\")\n",
    "        subname <- substr(subname, 1,(nchar(subname)-3))\n",
    "    }else if(subname == \"multi_seq_3_s1\"){\n",
    "        subdats <- c(\"multi_seq_3_s1\",\"multi_seq_3_s2\")\n",
    "        subname <- substr(subname, 1,(nchar(subname)-3))\n",
    "    }else if(subname == \"multi_4_allo1_1\"){\n",
    "        subdats <- c(\"multi_4_allo1_1\",\"multi_4_allo1_2\")\n",
    "        subname <- substr(subname, 1,(nchar(subname)-2))\n",
    "    }else if(subname == \"multi_bAllos_4_tum\"){\n",
    "        subdats <- c(\"multi_bAllos_4_PTs_S1\",\"multi_bAllos_4_PTs_S2\")\n",
    "    }else if(subname == \"multiseq_17\"){\n",
    "        subdats <- c(\"multiseq_17_S1\",\"multiseq_17_S2\")\n",
    "    }else{\n",
    "        subdats <- datasets[i]\n",
    "    }\n",
    "    \n",
    "    sub <- subset(dat, idents = subdats)\n",
    "        \n",
    "    allcells_eset <- t(as.matrix(GetAssayData(sub, slot = 'counts')))\n",
    "    \n",
    "    sub <- NormalizeData(sub)\n",
    "    all.genes <- rownames(x = sub)\n",
    "    sub <- ScaleData(object = sub, features = all.genes)\n",
    "    sub <- FindVariableFeatures(object = sub,selection.method = 'vst', nfeatures = 2000)\n",
    "    var_feats <- VariableFeatures(object = sub)\n",
    "    var_feats <- var_feats[!var_feats %in% c(\"mCherry\", \"EGFP\",\"MYCN\", \"LMO1\",\"dTomato\")]\n",
    "    var_feats <- var_feats[!var_feats %in% to.regress.cellcycle]\n",
    "    var_feats <- var_feats[!var_feats %in% regress_blood]\n",
    "\n",
    "    \n",
    "    write.table(allcells_eset, paste0(wd,\"./nmf_data/\",subname,\"_nrblstm_counts.tsv\"), quote = F, sep = \"\\t\")\n",
    "    write.table(var_feats, paste0(wd,\"./nmf_data/\",subname,\"_nrblstm_varGenes.txt\"), quote = F, sep = \"\\t\", row.names = F, col.names = F)\n",
    "    \n",
    "    rm(allcells_eset)\n",
    "    rm(sub)\n",
    "    rm(subname)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5111a18b-9fec-45f9-8354-3370b0e35ec4",
   "metadata": {},
   "source": [
    "## For individual tumor datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "302afb72-31b7-4755-9bae-c57836bbdb66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load genes associated with the cell cycle that should also be excluded for clustering\n",
    "to.regress.cellcycle <- read.csv(\"/data/junker/users/nfresma/zfish_prolif_markers.csv\")\n",
    "to.regress.cellcycle <- as.character(to.regress.cellcycle$x)\n",
    "\n",
    "# Get genes specific to blood cell types that are contaminants in NB cell data\n",
    "regress_blood <- c(\"hbba1\",\"hbba1.1\",\"hbae1\",\"hbbe2\",\"hbaa1\",\"hbaa2\",\"hbba2\",\"hbbe1.1\",\n",
    "                   \"lyz\",\"lect2l\",\"mmp9\",\"npsn\",\"srgn\",\n",
    "                   \"cd74a\",\"cd74b\",\"dusp2\",\n",
    "                   \"c1qb\",\"c1qc\",\"grn1\",\"marco\",\n",
    "                   \"tox\",\"zap70\",\"sla2\",\"igic1s1\",\n",
    "                   \"nkl.1\",\"nkl.4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3a5bab20-0245-4f32-ab03-9753d29db9b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Dataset EGFP_ctrl_1 is too small and will not be written to file for cNMF.\"\n",
      "[1] \"Dataset LMO1_ctrl_1 is too small and will not be written to file for cNMF.\"\n",
      "[1] \"Dataset LMO1_ctrl_m2 is too small and will not be written to file for cNMF.\"\n",
      "[1] \"Dataset MYCN_LMO1_vent_m7_1_1 is too small and will not be written to file for cNMF.\"\n",
      "[1] \"Dataset MYCN_LMO1_lat_m7_3_1 is too small and will not be written to file for cNMF.\"\n"
     ]
    }
   ],
   "source": [
    "# Per dataset:\n",
    "# Subset full dataset. Only keep processing data, if dataset contains at least 100 NB cells.\n",
    "# Find variable genes.\n",
    "# Remove known cell cycle markers, contaminating blood / immune cell genes and transgenes from variable genes list.\n",
    "# Write out count matrix and variable genes.\n",
    "\n",
    "Idents(dat) <- \"sample_all\"\n",
    "\n",
    "for(i in 1:length(unique(dat$sample_all))){\n",
    "    \n",
    "    \n",
    "    subname <- unique(dat$sample_all)[i]\n",
    "    sub <- subset(dat, idents = subname)\n",
    "    \n",
    "    \n",
    "    if(ncol(sub) < 100){\n",
    "        print(paste0('Dataset ',subname, ' is too small and will not be written to file for cNMF.'))\n",
    "        next\n",
    "    }\n",
    "    \n",
    "    allcells_eset <- t(as.matrix(GetAssayData(sub, slot = 'counts')))\n",
    "    \n",
    "    sub <- NormalizeData(sub, verbose = F)\n",
    "    all.genes <- rownames(x = sub)\n",
    "    sub <- FindVariableFeatures(object = sub,selection.method = 'vst', nfeatures = 2000, verbose = F)\n",
    "    \n",
    "    var_feats <- VariableFeatures(object = sub)\n",
    "    var_feats <- var_feats[!var_feats %in% c(\"mCherry\", \"EGFP\",\"MYCN\", \"LMO1\",\"dTomato\")]\n",
    "    var_feats <- var_feats[!var_feats %in% to.regress.cellcycle]\n",
    "    var_feats <- var_feats[!var_feats %in% to.regress]\n",
    "    var_feats <- var_feats[!var_feats %in% regress_blood]\n",
    "\n",
    "    \n",
    "    write.table(allcells_eset, paste0(wd, \"./nmf_data/\", subname, \"_nrblstm_counts.tsv\"), quote = F, sep = \"\\t\")\n",
    "    write.table(var_feats, paste0(wd, \"./nmf_data/\", subname, \"_nrblstm_varGenes.txt\"), quote = F, sep = \"\\t\", row.names = F, col.names = F)\n",
    "    \n",
    "    rm(allcells_eset)\n",
    "    rm(sub)\n",
    "    rm(subname)\n",
    "}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:r4]",
   "language": "R",
   "name": "conda-env-r4-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
